src/soc/ibm/power9/pstates.c

```diff
@@ -48,6 +53,117 @@ static ResonantClockingSetup resclk =
 	580		// L3 voltage threshold
 };
 
+#define WOF_IMAGE_MAGIC_VALUE (uint32_t)0x57544948 // "WTIH"
+#define WOF_IMAGE_VERSION     (uint32_t)1
+
+#define WOF_TABLES_MAGIC_VALUE (uint32_t)0x57465448 // "WFTH"
+#define WOF_TABLES_VERSION     (uint32_t)2
+#define WOF_TABLES_MAX_VERSION WOF_TABLES_VERSION 
+
+/*
+ * WOF image:
+ *  - header (struct wof_image_hdr)
+ *  - section table
+ *  - array of WOF tables
+ *
+ * WOF table:
+ *  - wof_tables_hdr for header
+ *  - data begins with vfrt_hdr
+ */
+
+/* Top-level header for WOF */
+struct wof_image_hdr
+{
+	uint32_t magic_number;	// WOF_IMAGE_MAGIC_VALUE
+	uint8_t  version;	// WOF_IMAGE_VERSION
+	uint8_t  entry_count;	// Number of entries in section table
+	uint32_t offset;	// BE offset to section table from image start
+} __attribute__((__packed__));
+
+/* Entry of WOF's section table */
+struct wof_image_entry
+{
+	uint32_t offset;	// BE offset to section from image start
+	uint32_t size;		// BE size of the section
+} __attribute__((__packed__));
+
+/* The values of wof_tables_hdr::mode */
+enum {
+	WOF_MODE_UNKNOWN = 0,	// Matches any value
+	WOF_MODE_NOMINAL = 1,
+	WOF_MODE_TURBO   = 2
+};
+
+/* Header of WOF's section */
+struct wof_tables_hdr
+{
+	uint32_t magic_number;	// WOF_TABLES_MAGIC_VALUE
+
+	struct
+	{
+		unsigned reserved: 20;
+		unsigned mode: 4;	// version 1: 0; version 2: 1 or 2; WOF_MODE_*
```

Comment:
Please drop the bitfields and use `uint16_t reserved; uint8_t mode;` instead. Bitfield endianness is undefined behavior, it works for GCC but may not work for other compilers.

After this change the inner struct won't be necessary.